payments-ui-core/src/main/java/com/stripe/android/ui/core/elements/CardNumberController.kt

```diff
@@ -86,18 +89,108 @@ internal class CardNumberEditableController constructor(
     // This makes the screen reader read out numbers digit by digit
     override val contentDescription: Flow<String> = _fieldValue.map { it.asIndividualDigits() }
 
-    override val cardBrandFlow = _fieldValue.map {
-        accountRangeService.accountRange?.brand ?: CardBrand.getCardBrands(it).firstOrNull()
+    private val isEligibleForCardBrandChoice = cardBrandChoiceConfig is CardBrandChoiceConfig.Eligible
+    private val brandChoices = MutableStateFlow<List<CardBrand>>(listOf())
+    private val chosenBrand = MutableStateFlow(
+        when (cardBrandChoiceConfig) {
+            is CardBrandChoiceConfig.Eligible -> {
+                when (val initialBrand = cardBrandChoiceConfig.initialBrand) {
+                    CardBrand.Unknown -> null
+                    else -> initialBrand
+                }
+            }
+            is CardBrandChoiceConfig.Ineligible -> null
+        }
+    )
+
+    /*
+     * In state validation, we check that the card number itself is valid and do not care about
+     * the card's co-brands. If a session is card brand choice eligible however, there is now the
+     * option  of not determining the card brand unless the user selects one. We use an implied
+     * card brand (VISA, Mastercard) internally to pass state validation.
+     */
+    private val impliedCardBrand = _fieldValue.map {
+        accountRangeService.accountRange?.brand
+            ?: CardBrand.getCardBrands(it).firstOrNull()
             ?: CardBrand.Unknown
     }
 
+    override val cardBrandFlow = impliedCardBrand.combine(
+        chosenBrand
+    ) { impliedBrand, chosenBrand ->
+        if (isEligibleForCardBrandChoice) {
+            chosenBrand ?: CardBrand.Unknown
+        } else {
+            impliedBrand
+        }
```

Comment:
This doesn’t cover all cases: If you enter `4242 4242`, you’d expect Visa to be displayed (because that card is not co-branded), but the field keeps showing the unknown brand icon.